File: packages/gavio-js/src/interceptors/audit/sink.ts
/** AuditSink — the extensible destination for audit records. */

import type { AuditRecord } from './record.js'

/** Where audit records go. Implement `write` to add a backend. */
export interface AuditSink {
  write(record: AuditRecord): Promise<void>
  /**
   * Erase records for a data subject (GDPR Art. 17, F-QUA-09). Remove every
   * persisted record whose `subjectId` matches and resolve to the number
   * removed. Optional — non-persistent sinks (e.g. stdout) omit it.
   */
  purge?(subjectId: string): Promise<number>
  /** Flush/close any resources. Optional. */
  close?(): Promise<void>
}

File: packages/gavio-py/gavio/interceptors/audit/sink.py
"""AuditSink — the extensible destination for audit records."""

from __future__ import annotations

from abc import ABC, abstractmethod

from .record import AuditRecord


class AuditSink(ABC):
    """Where audit records go. Implement ``write`` to add a backend."""

    @abstractmethod
    async def write(self, record: AuditRecord) -> None: ...

    async def purge(self, subject_id: str) -> int:
        """Erase records for a data subject (GDPR Art. 17, F-QUA-09).

        Remove every persisted record whose ``subject_id`` matches and return
        the number removed. The default is a no-op returning 0 — appropriate for
        non-persistent sinks (e.g. stdout). Persistent sinks override this.
        """
        return 0

    async def close(self) -> None:
        """Flush/close any resources. Default no-op."""
        return None
